executor/index_merge_reader.go

```diff
@@ -840,6 +872,194 @@ type indexMergeProcessWorker struct {
 	stats      *IndexMergeRuntimeStat
 }
 
+type rowIdx struct {
+	partialID int
+	taskID    int
+	rowID     int
+}
+
+type handleHeap struct {
+	requiredCnt uint64
+	taskMap     map[int][]*indexMergeTableTask
+
+	idx         []rowIdx
+	compareFunc []chunk.CompareFunc
+	byItems     []*plannerutil.ByItems
+}
+
+func (h handleHeap) Len() int {
+	return len(h.idx)
+}
+
+func (h handleHeap) Less(i, j int) bool {
+	rowI := h.taskMap[h.idx[i].partialID][h.idx[i].taskID].idxRows.GetRow(h.idx[i].rowID)
+	rowJ := h.taskMap[h.idx[j].partialID][h.idx[j].taskID].idxRows.GetRow(h.idx[j].rowID)
+
+	for i, compFunc := range h.compareFunc {
+		cmp := compFunc(rowI, i, rowJ, i)
+		if !h.byItems[i].Desc {
+			cmp = -cmp
+		}
+		if cmp < 0 {
+			return true
+		} else if cmp > 0 {
+			return false
+		}
+	}
+	return false
+}
+
+func (h handleHeap) Swap(i, j int) {
+	h.idx[i], h.idx[j] = h.idx[j], h.idx[i]
+}
+
+func (h *handleHeap) Push(x interface{}) {
+	idx := x.(rowIdx)
+	h.idx = append(h.idx, idx)
+}
+
+func (h *handleHeap) Pop() interface{} {
+	idxRet := h.idx[len(h.idx)-1]
+	h.idx = h.idx[:len(h.idx)-1]
+	return idxRet
+}
+
+func (w *indexMergeProcessWorker) NewHandleHeap(taskMap map[int][]*indexMergeTableTask) *handleHeap {
+	compareFuncs := make([]chunk.CompareFunc, 0, len(w.indexMerge.byItems))
+	for _, item := range w.indexMerge.byItems {
+		keyType := item.Expr.GetType()
+		compareFuncs = append(compareFuncs, chunk.GetCompareFunc(keyType))
+	}
+	requiredCnt := w.indexMerge.pushedLimit.Count + w.indexMerge.pushedLimit.Offset
+	return &handleHeap{
+		requiredCnt: requiredCnt,
+		taskMap:     taskMap,
+		idx:         make([]rowIdx, 0, requiredCnt),
+		compareFunc: compareFuncs,
+		byItems:     w.indexMerge.byItems,
+	}
+}
+
+// pruneTableWorkerTaskIdxRows prune idxRows and only keep columns that will be used in byItems.
+func (w *indexMergeProcessWorker) pruneTableWorkerTaskIdxRows(task *indexMergeTableTask) {
+	// IndexScan no need to prune retChk, Columns required by byItems are always first.
+	if plan, ok := w.indexMerge.partialPlans[task.partialPlanID][0].(*plannercore.PhysicalTableScan); ok {
+		prune := make([]int, 0, len(w.indexMerge.byItems))
+		for _, item := range plan.ByItems {
+			c, _ := item.Expr.(*expression.Column)
```

Comment:
The `item.Expr.(*expression.Column).Index` is the same thing as the `i` in `for i, col := range plan.Schema().Columns`.
You can use the `Index` directly.

Reply:
Updated.
It confused me about `ID` and `UniqueID` in `expression.Column`. I found we use `ID` to compare with  `model.ColumnInfo`, and use `UniqueID` to compare with `expression.Column`

Reply:
Yes, ID is the meta data of `model.ColumnInfo`, created by DDL. So `ID` is a table-level meta data.
`UniqueID` is generated during planning. It's unique at the SQL level.
We often use `UniqueID` to check the equivalence of the `expression.Column`. Use `ID` to know which `ColumnInfo` it represents.